File: test/extract_wiktextract_test.py
import json
import os

LEXICON_PATH = "../dic/lexicon.jsonl"
WIKTEXTRACT_PATH = "../tmp/raw-wiktextract-data.jsonl"
OUTPUT_PATH = "../tmp/wiktextract_test_data.jsonl"

def load_lexicon_phrases():
    phrases = set()
    with open(LEXICON_PATH, 'r') as f:
        for line in f:
            if not line.strip():
                continue
            entry = json.loads(line)
            phrase = entry["phrase"].lower()
            if " " in phrase or "-" in phrase:
                phrases.add(phrase)
    return phrases

def extract_test_data(lexicon_phrases):
    print("Extracting test sentences from wiktextract...")

    if not os.path.exists(WIKTEXTRACT_PATH):
        print(f"Error: {WIKTEXTRACT_PATH} not found.")
        return

    count = 0
    with open(WIKTEXTRACT_PATH, 'r') as f_in, open(OUTPUT_PATH, 'w') as f_out:
        for line in f_in:
            if not line.strip():
                continue
            try:
                entry = json.loads(line)
            except json.JSONDecodeError:
                continue

            word = entry.get("word", "").lower()

            # Only care about multi-word phrases that are in our lexicon
            if word in lexicon_phrases:
                for sense in entry.get("senses", []):
                    for example in sense.get("examples", []):
                        if "text" in example:
                            text = example["text"].strip()
                            # Basic length filtering to avoid absurdly long or short examples
                            if 15 < len(text) < 300:
                                json.dump({"text": text, "expected_phrase": word}, f_out)
                                f_out.write("\n")
                                count += 1

    print(f"Extracted and saved {count} test sentences to {OUTPUT_PATH}.")

if __name__ == "__main__":
    print("Loading Lexicon...")
    lexicon_phrases = load_lexicon_phrases()
    print(f"Loaded {len(lexicon_phrases)} multi-word phrases from lexicon.")
    extract_test_data(lexicon_phrases)

File: src/bin/extract_wiktextract_test.rs
use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::fs::File;
use std::io::{BufRead, BufReader, BufWriter, Write};

#[derive(Deserialize)]
struct LexiconEntry {
    phrase: String,
}

#[derive(Deserialize)]
struct WiktExample {
    text: Option<String>,
}

#[derive(Deserialize)]
struct WiktSense {
    #[serde(default)]
    examples: Vec<WiktExample>,
}

#[derive(Deserialize)]
struct WiktEntry {
    #[serde(default)]
    word: String,
    #[serde(default)]
    senses: Vec<WiktSense>,
}

#[derive(Serialize)]
struct OutputEntry<'a> {
    text: &'a str,
    expected_phrase: &'a str,
}

fn main() -> Result<()> {
    let lexicon_path = "dic/lexicon.jsonl";
    let wiktextract_path = "tmp/raw-wiktextract-data.jsonl";
    let output_path = "tmp/wiktextract_test_data.jsonl";

    println!("Loading Lexicon...");
    let mut lexicon_phrases = HashSet::new();
    let lex_file = File::open(lexicon_path).context("Failed to open lexicon")?;
    for line in BufReader::new(lex_file).lines() {
        let line = line?;
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        if let Ok(entry) = serde_json::from_str::<LexiconEntry>(line) {
            let phrase = entry.phrase.to_lowercase();
            if phrase.contains(' ') || phrase.contains('-') {
                lexicon_phrases.insert(phrase);
            }
        }
    }
    println!(
        "Loaded {} multi-word phrases from lexicon.",
        lexicon_phrases.len()
    );

    println!("Extracting test sentences from wiktextract...");
    if !std::path::Path::new(wiktextract_path).exists() {
        println!("Error: {} not found.", wiktextract_path);
        return Ok(());
    }

    let wikt_file = File::open(wiktextract_path).context("Failed to open wiktextract data")?;
    let out_file = File::create(output_path).context("Failed to create output file")?;
    let mut out_writer = BufWriter::new(out_file);
    let mut count = 0;

    for line in BufReader::new(wikt_file).lines() {
        let line = line?;
        let line = line.trim();
        if line.is_empty() {
            continue;
        }

        let Ok(entry) = serde_json::from_str::<WiktEntry>(line) else {
            continue;
        };

        let word = entry.word.to_lowercase();
        if lexicon_phrases.contains(&word) {
            for sense in entry.senses {
                for example in sense.examples {
                    if let Some(text) = example.text {
                        let text = text.trim();
                        let char_count = text.chars().count();
                        if char_count > 15 && char_count < 300 {
                            let out_entry = OutputEntry {
                                text,
                                expected_phrase: &word,
                            };
                            serde_json::to_writer(&mut out_writer, &out_entry)?;
                            out_writer.write_all(b"\n")?;
                            count += 1;
                        }
                    }
                }
            }
        }
    }

    out_writer.flush()?;
    println!(
        "Extracted and saved {} test sentences to {}.",
        count, output_path
    );

    Ok(())
}
